pkg/runcfanotify/runcfanotify.go

```diff
@@ -0,0 +1,369 @@
+	return strings.Split(string(cmdline), "\x00")
+}
+
+func (n *RuncNotifier) watchPidFileIterate(pidFileDirNotify *fanotify.NotifyFD, bundleDir string, pidFile string, pidFileDir string) (bool, error) {
+	// Get the next event from fanotify.
+	// Even though the API allows to pass skipPIDs, we cannot use
+	// it here because ResponseAllow would not be called.
+	data, err := pidFileDirNotify.GetEvent()
+	if err != nil {
+		return false, fmt.Errorf("%w", err)
+	}
+
+	// data can be nil if the event received is from a process in skipPIDs.
+	// In that case, skip and get the next event.
+	if data == nil {
+		return false, nil
+	}
+
+	// Don't leak the fd received by GetEvent
+	defer data.Close()
+
+	if !data.MatchMask(unix.FAN_ACCESS_PERM) {
+		// This should not happen: FAN_ACCESS_PERM is the only mask Marked
+		return false, fmt.Errorf("fanotify: unknown event on runc: mask=%d pid=%d", data.Mask, data.Pid)
+	}
+
+	// This unblocks whoever is accessing the pidfile
+	defer pidFileDirNotify.ResponseAllow(data)
+
+	pid := data.GetPID()
+
+	// Skip events triggered by ourselves
+	if pid == os.Getpid() {
+		return false, nil
+	}
+
+	path, err := data.GetPath()
+	if err != nil {
+		return false, err
+	}
+	if path != pidFile {
+		return false, nil
+	}
+
+	pidFileContent, err := ioutil.ReadAll(data.File())
+	if err != nil {
+		return false, err
+	}
+	if len(pidFileContent) == 0 {
+		return false, fmt.Errorf("empty pid file")
+	}
+	containerPID, err := strconv.Atoi(string(pidFileContent))
+	if err != nil {
+		return false, err
+	}
+
+	// Unfortunately, Linux 5.4 doesn't respect ignore masks
+	// See fix in Linux 5.9:
+	// https://github.com/torvalds/linux/commit/497b0c5a7c0688c1b100a9c2e267337f677c198e
+	// Workaround: remove parent mask. We don't need it anymore :)
+	err = pidFileDirNotify.Mark(unix.FAN_MARK_REMOVE, unix.FAN_ACCESS_PERM|unix.FAN_EVENT_ON_CHILD, unix.AT_FDCWD, pidFileDir)
+	if err != nil {
+		return false, nil
+	}
+
+	pidfd, _, errno := unix.Syscall(unix.SYS_PIDFD_OPEN, uintptr(containerPID), 0, 0)
+	if errno != 0 {
+		return false, fmt.Errorf("pidfd_open returned %v", errno)
+	}
+
+	bundleConfigJson, err := ioutil.ReadFile(filepath.Join(bundleDir, "config.json"))
+	if err != nil {
+		return false, err
+	}
+	containerConfig := &ocispec.Spec{}
+	err = json.Unmarshal(bundleConfigJson, containerConfig)
+	if err != nil {
+		return false, err
+	}
+
+	containerID := filepath.Base(filepath.Clean(bundleDir))
+
+	go func() {
+		defer unix.Close(int(pidfd))
+		for {
+			fds := []unix.PollFd{
+				unix.PollFd{
+					Fd:      int32(pidfd),
+					Events:  unix.POLLIN,
+					Revents: 0,
+				},
+			}
+			count, err := unix.Poll(fds, -1)
+			if err == nil && count == 1 {
+				n.callback(ContainerEvent{
+					Type:            EVENT_TYPE_REMOVE_CONTAINER,
+					ContainerID:     containerID,
+					ContainerPID:    uint32(containerPID),
+					ContainerConfig: containerConfig,
+				})
+				return
+			}
+		}
```

Comment:
comment for another PR. Perhaps we could have a single goroutine to monitor all containers?